[Win] Add an alternative Skia backend support

fujii opened the pull request:
#### 5bd5f37115b2e0314de307f68629d2b1a92791f6
<pre>
[Win] Add an alternative Skia backend support
<a href="https://bugs.webkit.org/show_bug.cgi?id=269117">https://bugs.webkit.org/show_bug.cgi?id=269117</a>

Reviewed by Carlos Garcia Campos.

Build out the bundled Skia. Harfbuzz is required.

&gt; build-webkit --cmakeargs=&quot;-DUSE_SKIA=1&quot;

There are still some problems.
* custom fonts don&apos;t work.
* pixel formats of video isn&apos;t correct
* linked font doesn&apos;t work.
* Not implemented yet: GraphicsContext::releaseWindowsContext, DragImageWin.cpp and ImageAdapterWin.cpp

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformDataAttributes::FontPlatformDataAttributes):
(WebCore::FontPlatformData::hfont const):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::platformInit):
(WebCore::FontCache::fontManager const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::updateContents):
* Source/WebCore/platform/graphics/win/FontPlatformDataWin.cpp:
(WebCore::FontPlatformData::FontPlatformData):
* Source/WebCore/platform/graphics/win/GraphicsContextWin.cpp:
(WebCore::GraphicsContext::releaseWindowsContext):
* Source/WebCore/platform/graphics/win/ImageAdapterWin.cpp:
(WebCore::ImageAdapter::getHBITMAPOfSize):
(WebCore::ImageAdapter::nativeImageOfHBITMAP):
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
(WebCore::MediaPlayerPrivateMediaFoundation::Direct3DPresenter::paintCurrentFrame):
* Source/WebCore/platform/graphics/win/cairo/FontPlatformDataWinCairo.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::platformIsEqual const):
(WebCore::FontPlatformData::familyName const):
* Source/WebCore/platform/win/DragImageWin.cpp:
(WebCore::createDragImageForLink):
(WebCore::createDragImageFromImage):
(WebCore::scaleDragImage):
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/skia/Skia.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder&lt;SkString&gt;::encode):
(IPC::ArgumentCoder&lt;SkString&gt;::decode):
(IPC::ArgumentCoder&lt;SkFontStyle::Slant&gt;::encode):
(IPC::ArgumentCoder&lt;SkFontStyle::Slant&gt;::decode):
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
(WebKit::BackingStore::paint):
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.cpp:
(WebKit::WebPopupMenuProxyWin::paint):
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.h:
* Source/WebKit/UIProcess/win/WebView.cpp:
(WebKit::WebView::paint):
* Source/cmake/OptionsWin.cmake:
* Tools/ImageDiff/PlatformWin.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/win/LinkedFonts.cpp:
* Tools/WebKitTestRunner/PlatformWin.cmake:
* Tools/WebKitTestRunner/win/PlatformWebViewWin.cpp:
(WTR::generateCairoSurfaceFromBitmap):
(WTR::PlatformWebView::windowSnapshotImage):

Canonical link: <a href="https://commits.webkit.org/286632@main">https://commits.webkit.org/286632@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc99e131769e731607eccbdb6abbb08d91080bc5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81117 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27863 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78704 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64764 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60043 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18142 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23251 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26187 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69770 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68480 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82563 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75864 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3963 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2608 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68321 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4116 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65730 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67570 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11538 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98118 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11844 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3910 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6719 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21463 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3933 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7363 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5691 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont reviewed (commented):
Overall this is fine in terms of the build but have one large issue with it

We should not be hand rolling any serialization code and land https://github.com/WebKit/WebKit/pull/27902 . Can you check to see if it works for you?

Other thing is do we really need harfbuzz? If we can avoid adding it in that would be ideal.

carlosgcampos commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -905,18 +906,48 @@ if (ANDROID)
     target_sources(Skia PRIVATE src/ports/SkDebug_stdio.cpp)
     target_sources(Skia PRIVATE src/ports/SkFontMgr_fontconfig.cpp)
+    target_sources(Skia PRIVATE
+	src/ports/SkOSFile_posix.cpp
```
```suggestion
	    src/ports/SkOSFile_posix.cpp
```

fujii replied:
Fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -209,6 +216,11 @@ void BitmapTexture::updateContents(NativeImage* frameImage, const IntRect& targe
     int bytesPerLine = cairo_image_surface_get_stride(surface);
 
     updateContents(imageData, targetRect, offset, bytesPerLine, PixelFormat::BGRA8);
+#elif USE(SKIA)
+    sk_sp<SkImage> surface = frameImage->platformImage();
+    SkPixmap pixmap;
+    if (surface->peekPixels(&pixmap))
+        updateContents(pixmap.addr(), targetRect, offset, pixmap.rowBytes(), PixelFormat::BGRA8);
```
Is this SkImage always unaccelerated?

fujii replied:
Yes, at the moment.
https://github.com/WebKit/WebKit/blob/1b0ca739c27ecccb811e64e46232f33b1401c5ca/Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp#L221-L223
`ImageBufferOptions::Accelerated` has to be specified.

carlosgcampos commented on Source/WebKit/Shared/skia/Skia.serialization.in:
```diff
@@ -0,0 +1,28 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+header: <skia/core/SkFontStyle.h>
+[CustomHeader] class SkFontStyle {
+    int weight();
+    int width();
+    SkFontStyle::Slant slant();
+};
```
Can't we add this to WebCoreFont with a skia ifdef? I'm not sure we want a generic Skia file for serializations, I think it's better to have different files for different things.

fujii replied:
I think all skia types `SkString`, `SkFontStyle::Slant`, `SkColorSpace`, `SkData` can be moved to this `Skia.serialization.in`.

achristensen07 commented on Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
```diff
@@ -32,9 +32,41 @@
 
 namespace IPC {
 
+void ArgumentCoder<SkString>::encode(Encoder& encoder, const SkString& string)
```
Please use serialization.in files instead of this.

fujii replied:
We need more documents to migrate. It took hours to me to find out the undocumented "wrapped by" keyword.

fujii replied:
`wrapped by` is only for ObjC type. I can't use it for Skia types now. Needs to modify generate-serializers.py. Will try in a follow up patch.

fujii committed:
[Win] Add an alternative Skia backend support
https://bugs.webkit.org/show_bug.cgi?id=269117

Reviewed by Carlos Garcia Campos.

Build out the bundled Skia. Harfbuzz is required.

> build-webkit --cmakeargs="-DUSE_SKIA=1"

There are still some problems.
* custom fonts don't work.
* pixel formats of video isn't correct
* linked font doesn't work.
* Not implemented yet: GraphicsContext::releaseWindowsContext, DragImageWin.cpp and ImageAdapterWin.cpp

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformDataAttributes::FontPlatformDataAttributes):
(WebCore::FontPlatformData::hfont const):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::platformInit):
(WebCore::FontCache::fontManager const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::updateContents):
* Source/WebCore/platform/graphics/win/FontPlatformDataWin.cpp:
(WebCore::FontPlatformData::FontPlatformData):
* Source/WebCore/platform/graphics/win/GraphicsContextWin.cpp:
(WebCore::GraphicsContext::releaseWindowsContext):
* Source/WebCore/platform/graphics/win/ImageAdapterWin.cpp:
(WebCore::ImageAdapter::getHBITMAPOfSize):
(WebCore::ImageAdapter::nativeImageOfHBITMAP):
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
(WebCore::MediaPlayerPrivateMediaFoundation::Direct3DPresenter::paintCurrentFrame):
* Source/WebCore/platform/graphics/win/cairo/FontPlatformDataWinCairo.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::platformIsEqual const):
(WebCore::FontPlatformData::familyName const):
* Source/WebCore/platform/win/DragImageWin.cpp:
(WebCore::createDragImageForLink):
(WebCore::createDragImageFromImage):
(WebCore::scaleDragImage):
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/skia/Skia.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder<SkString>::encode):
(IPC::ArgumentCoder<SkString>::decode):
(IPC::ArgumentCoder<SkFontStyle::Slant>::encode):
(IPC::ArgumentCoder<SkFontStyle::Slant>::decode):
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
(WebKit::BackingStore::paint):
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.cpp:
(WebKit::WebPopupMenuProxyWin::paint):
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.h:
* Source/WebKit/UIProcess/win/WebView.cpp:
(WebKit::WebView::paint):
* Source/cmake/OptionsWin.cmake:
* Tools/ImageDiff/PlatformWin.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/win/LinkedFonts.cpp:
* Tools/WebKitTestRunner/PlatformWin.cmake:
* Tools/WebKitTestRunner/win/PlatformWebViewWin.cpp:
(WTR::generateCairoSurfaceFromBitmap):
(WTR::PlatformWebView::windowSnapshotImage):

Canonical link: https://commits.webkit.org/286632@main
